### ISSUE TITLE
Message Requests + Discovery Context

### ISSUE THREAD

**Issue by ShadowJonathan** [Yellow-Dog-Man/Resonite-Issues#943]
### Is your feature request related to a problem? Please describe.

Currently, when users are in a world, they have to send contact requests before they can open a dialog with that user.

This adds a little bit of friction, which may be unnecessary when the user just wishes to "whisper" something at a local time, without needing to add as a contact.

Additionally, someone may want to send a message to a creator, without adding them as a contact, for the purpose of it being just a short-lived interaction, and not much more.

If this were added as-is, it would be hard to figure out when and where the user would've discovered the other user, which would be useful in providing awareness context of "where they might've gotten picked up"

### Describe the solution you'd like

A "Message Request" inbox where users can receive incoming messages from users that they haven't "met" before, which then would be hidden, out of sight, or have independent/recluded number badges, to decrease the impact or overwhelmingness of spam.

Additionally, a feature where a user on a local world can directly send the other person a message (that then appears in the message box list) that gets the normal sounds and badges. This way, the user can be contacted non-verbally, or "whispered to", without needing to be a lasting contact.

This latter one would be accompanied with a DM starting badge (or other badge) saying "You have met this user on X world at Y date-time". A "close" button would be provided", and when the user exits the world, this DM would migrate to Message Requests.

(Additionally, for safety and privacy reasons, it might also be possible to enable/disable by-default the "allow DMs from world users", and/or change for a particular world (for a particular event). It might also be useful to have this setting depend on session size, i.e. "any world with 10 or more users should have DMs go to requests")

#### Keep the contacts system

While it might be useful, it might bring undue friction or otherwise bloat that contacts list to be something "less" than friends.

For this, I'd suggest to add a friends, "favourite contacts" list, or "contact groups", to mitigate this.

#### Just add the Message Requests, not the world DMs

This is a viable option, and might be better, since the World/Session DMs adds additional security attack surface.

#### Add a "session chat"

(Not too sure if this already exists)

It might be possible to circumvent this requirement entirely and just add a "Session chat", where users see chat messages from different worlds they're logged into, at the same time, so that then anyone they find in there can be "upgraded" to contact.

This chat would work similarly a-la minecraft chat; There is a "broadcast" chat setting, and then there's a "`/msg`"/"whisper" command to send a specific user a specific message via this session chat.

This might become useful when, for example, someone is in another session while still logged into another session, and they're urgently needed (while they dont have contacts added there), so someone can whisper some information to them via this chat.

This chat would be ephemeral, and be removed when the host closes the session.

**Comment by ShadowJonathan** [comment]
(After thinking about it, I'm beginning to be a bit more partial of the idea of Message Requests + Session Chat, since the idea of "allowing anyone to just send a DM with a vague 'you met this user here' marker" might be a bit too much a moderation/security nightmare, and be too exploitable)

**Comment by Banane9** [comment]
Some user tools like the mute helper already offer a sort of session chat log - it would be nice if that eventually became a first party feature 🤔 

One of the funniest things of the current contact messaging system is also, that you can send a message to anyone who added *you* as a contact - it doesn't have to be mutual. So if you get a request you can send them messages they can't reply to.

**Comment by shiftyscales** [comment]
In a different issue, I'd made a [comment](https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/1508#issuecomment-2067389143) that covers this functionality, and more broadly- chat system mechanics/considerations. It can be seen in full at the bottom of this post for further context.

The "discovery context" portion of your issue would be covered under another issue, #1108.

I'm updating the title to focus on the primary request of this issue 'message requests'. Other functionality mentioned such as having an in-world chat system was requested in #50.

Moreover- this issue is going to be blocked by the contacts UI rework #1116, as we are not going to be prioritizing further development to legacy UIs like the contacts UI.

> > But there's other issues too - can you bring in users who are not your contacts? Can any contacts bring other contacts in there? Or do all members need to be contacts with each other?
> 
> Using Discord as a point of reference- as far as I'm aware any member of a chat can add any friend of theirs to the chat. **However**, I don't think that is a good solution, and we'd be better to implement our own logic.
> 
> - This should be something that could be set as a privacy setting on a per-chat basis
> - E.g. by default only the creator of the group chat could add new members to it, unless they explicitly grant permission to other members of the group chat, either on a per-user or global basis
> - This would avoid situations in which a potentially malicious (or accidental) invite exposes the contents of a chat when it is undesired.
> - E.g. if someone were talking in a group chat about planning a surprise birthday party for a user, and someone in the group added the user it is about to the group.
> 
> ---
> 
> Likewise- I think it would desirable for users to have a privacy setting that determines who is allowed to invite them into a group chat/contact them.
> - E.g. only contacts by default could invite them to a group chat, but they could optionally set it to everyone/no-one
> - This would prevent solicitation/maliciously pulling users into group chats they didn't consent to be added to
> - It may be desirable to make group chats a type of 'invite' that way, if they are invited to something it is ultimately their choice whether or not they accept that invite regardless of how they have their settings configured
> 
> ---
> 
> In Discord, there are 'message request' settings, which allow messages from non-contacts to be set aside in their own inbox. This is configurable on a per-server basis (or can be disabled entirely to prevent DMs from server (group) members
> - This functionality could be considered as a privacy setting along with the above suggestions whether the user wishes to receive message requests from non-contacts/group members they are not contacts with
> - By default we should only allow users to receive messages from contacts/this functionality should be disabled by default
> - We do not want to allow any user to be able to message any other user without a user consenting/opting in to receiving message requests from non-contacts or group members
> 
> Likewise- Discord also has privacy settings for who can add the user as a contact, e.g. everyone, contacts of contacts, or nobody (the user can only send outbound contact requests, not receive them).
> - This would need to be set to `everyone` by default by necessity- but would be an important privacy setting, particularly for content creators, and users who do not wish to be able to be added as a contact by any user on the platform
> 
> ---
> 
> Overall, when establishing guidelines/rules/mechanics, we will want to ensure that the chat systems are built in such a way to minimize malicious activity, e.g. solicitation, spam, unsolicited/inappropriate messaging, etc.
> 
> The current platform defaults of only being able to receive messages from contacts is a good starting point to build off of, so likewise- receiving a group chat invite should be something that is contacts-only by default, and for the creator of the group chat- should be 'private' unless they choose to lessen the restrictions.
> 
> Effectively:
> - **Treat it in a similar way to how our session access works**
> - **Give users the choice of whether or not to accept the group chat invite rather than automatically bringing them into it**

**Comment by Amunak** [comment]
I'm not sure messages should be disabled by default. There should be an option to just block (messages from) people, but you should be able to message anyone, really, with potentially having a settings option to allow only users in the worlds you are in message you, or disable messages from non-friends completely.

But currently it's extremely inconvenient; especially when a new user joins and the session (or the user) crashes / get yeeted for whatever reason, they might not know what happened and it's hard to get them back. Being able to send a quick message and an invite would help a lot in such cases.
